src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,668 @@
-                    new_bindings
-                        .iter()
-                        .map(|(partition, diff)| ((&(), *partition), &new_ts, diff)),
-                    self.write_upper.clone(),
-                    new_upper.clone(),
-                )
-                .await
-                .expect("Timestamper CAA")
-                .expect("Timestamper CAA 2");
-
-            // We don't just return here on success because the logic to determine what progress to return is complex
-            // and should be in one place only.
-            self.write_upper = match compare_and_append_result {
-                Ok(()) => new_upper,
-                Err(Upper(actual_upper)) => actual_upper,
-            };
+        }
 
-            while PartialOrder::less_than(&self.read_progress, &self.write_upper) {
+        // If there no updates we might still want to bump the upper time if enough time has passed
+        let mint_timestamp = if updates.is_empty() {
+            self.next_mint_timestamp().now_or_never()
+        } else {
+            Some(self.next_mint_timestamp().await)
+        };
+
+        if let Some(next_ts) = mint_timestamp {
+            let new_upper = Antichain::from_elem(next_ts + 1);
+            loop {
+                let upper = self.upper.clone();
+                let new_upper = new_upper.clone();
+                let updates = updates.iter().map(|(pid, diff)| (((), pid), next_ts, diff));
                 match self
-                    .timestamp_bindings_listener
-                    .next()
+                    .write_handle
+                    .compare_and_append(updates, upper, new_upper)
                     .await
-                    .expect("ListenStream doesn't end")
                 {
-                    ListenEvent::Progress(progress) => {
-                        assert!(
-                            timely::PartialOrder::less_equal(&self.read_progress, &progress),
-                            "{:?} PARTIAL ORDER: {:?} {:?}",
-                            self.name,
-                            self.read_progress,
-                            progress
-                        );
-                        self.read_progress = progress;
-                    }
-                    ListenEvent::Updates(updates) => {
-                        for ((_, value), timestamp, diff) in updates {
-                            let partition = value.expect("Unable to decode partition id");
-                            self.persisted_timestamp_bindings
-                                .entry(partition)
-                                .or_insert_with(VecDeque::new)
-                                .push_back((timestamp, diff));
-                        }
-                    }
+                    Ok(Ok(Ok(()))) => break,
+                    Ok(Ok(Err(actual_upper))) => return Err(actual_upper),
+                    Ok(Err(invalid_use)) => panic!("{invalid_use:?}"),
+                    Err(_external_err) => continue,
```

Comment:
`tracing::debug` the error at least

also, can we add a comment on why retrying is okay?

Reply:
I added a comment explaining why retrying is ok. Correctness isn't relying on any particular pattern of calling `mint` or `sync`. You can call them however you want (mint without sync, sync to wrong frontiers then mint, etc) and you'll never get back bad bindings. You'll either get correctly minted bindings or an error or a hang (e.g if you sync past the upper).

Calling the methods in the way they are in this PR just ensures that we make forward progress